server/services/v1/database/query_index.go

```diff
@@ -0,0 +1,294 @@
+	"context"
+	"fmt"
+
+	"github.com/buger/jsonparser"
+	"github.com/rs/zerolog/log"
+	"github.com/tigrisdata/tigris/keys"
+	"github.com/tigrisdata/tigris/schema"
+	"github.com/tigrisdata/tigris/server/transaction"
+	"github.com/tigrisdata/tigris/store/kv"
+	"github.com/tigrisdata/tigris/value"
+	"golang.org/x/text/collate"
+	"golang.org/x/text/language"
+)
+
+const (
+	MAX_STRING_LEN = 64
+)
+
+type IndexIterator struct {
+	tx   transaction.Tx
+	it   kv.Iterator
+	ctx  context.Context
+	keys []keys.Key
+	err  error
+}
+
+type IndexRow struct {
+	docId []byte
+}
+
+func (iter *IndexIterator) Next(row *kv.KeyValue) bool {
+	if iter.err != nil {
+		return false
+	}
+
+	if iter.it.Next(row) {
+		return true
+	}
+
+	return false
+}
+
+func (it *IndexIterator) Interrupted() error { return it.err }
+
+type FieldIndex struct {
+	value value.Value
+	name  string
+	pos   int
+}
+
+func newFieldIndex(dataType schema.FieldType, name string, rawValue []byte, pos int) (*FieldIndex, error) {
+	if dataType == schema.StringType {
+		rawValue = collateString(rawValue)
+	}
+
+	value, err := value.NewValue(dataType, rawValue)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &FieldIndex{
+		value,
+		name,
+		pos,
+	}, nil
+}
+
+type QueryIndexInfo struct {
+	Rows int64
+	Size int64
+}
+
+func genRowCountKey(coll *schema.DefaultCollection) keys.Key {
+	return keys.NewKey(coll.EncodedName, coll.Indexes.QueryIndex.Name, "_info", "count")
+}
+
+func genIndexSizeKey(coll *schema.DefaultCollection) keys.Key {
+	return keys.NewKey(coll.EncodedName, coll.Indexes.QueryIndex.Name, "_info", "size")
+}
+
+type QueryIndex struct {
+	objectDepth int
+}
+
+func NewQueryIndexStore() *QueryIndex {
+
+	return &QueryIndex{
+		objectDepth: 4,
+	}
+}
+
+func (q *QueryIndex) ScanIndex(ctx context.Context, tx transaction.Tx, coll *schema.DefaultCollection, start keys.Key, end keys.Key) (*IndexIterator, error) {
+
+	it, err := tx.ReadRange(ctx, start, end, false)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &IndexIterator{
+		tx:   tx,
+		it:   it,
+		keys: []keys.Key{start, end},
+	}, nil
+}
+
+func (q *QueryIndex) IndexInfo(ctx context.Context, tx transaction.Tx, coll *schema.DefaultCollection) (*QueryIndexInfo, error) {
+	rowCountKey := genRowCountKey(coll)
+	indexSizeKey := genIndexSizeKey(coll)
+
+	rows, err := tx.AtomicRead(ctx, rowCountKey)
+
+	if err != nil {
+		return nil, err
+	}
+
+	size, err := tx.AtomicRead(ctx, indexSizeKey)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &QueryIndexInfo{
+		rows,
+		size,
+	}, nil
+}
+
+func (q *QueryIndex) Index(ctx context.Context, tx transaction.Tx, coll *schema.DefaultCollection, doc []byte, id []interface{}) error {
+
+	indexKeys, size, rowCount := q.buildKVs(coll, doc, id)
+
+	rowCountKey := genRowCountKey(coll)
+	indexSizeKey := genIndexSizeKey(coll)
+
+	tx.AtomicAdd(ctx, rowCountKey, rowCount)
+	tx.AtomicAdd(ctx, indexSizeKey, size)
+
+	for _, indexKey := range indexKeys {
+		err := tx.Insert(ctx, indexKey, nil)
```

Comment:
Also, if the user is using insert API, then blindly replacing an index is good enough. In the case of replace/update, we also need the old row to clean up old indexes.